pkg/router/template/router.go

```diff
@@ -1369,10 +1369,16 @@ func (r *templateRouter) getActiveEndpoints(serviceUnits map[ServiceUnitKey]int3
 func (r *templateRouter) calculateServiceWeights(serviceUnits map[ServiceUnitKey]int32, port string) map[ServiceUnitKey]int32 {
 	serviceUnitNames := make(map[ServiceUnitKey]int32)
 
-	// If there is only 1 service unit, then always set the weight 1
+	// If there is only 1 active service unit, then always set the weight 1
 	// for all the endpoints, except when the service weight is 0.
 	// Scaling the weight to 256 is redundant and causes haproxy to allocate more memory on startup.
-	if len(serviceUnits) == 1 {
+	activeServiceUnits := 0
+	for key, weight := range serviceUnits {
+		if weight > 0 && r.numberOfEndpoints(key, port) > 0 {
+			activeServiceUnits++
+		}
+	}
+	if activeServiceUnits == 1 {
```

Comment:
Does this work the same, without counting active services ahead of time?

> for key, weight := range serviceUnits {
> 		if weight > 0 && r.numberOfEndpoints(key, port) > 0 {
> 			serviceUnitNames[key] = 1
> 		} else {
> 			serviceUnitNames[key] = 0
> 		}
> 	}

Reply:
It doesn't - you need to count the total number of active endpoints first to determine if you should force weight to be 1.

It's not clear if you meant to keep `if len(serviceUnits) == 1 {` in your suggested logic. I'll assume you meant this because otherwise your suggestion would set all active services to weight of 1, which would defeat the purpose of weight scaling entirely:
```
if len(serviceUnits) == 1 {
    for key, weight := range serviceUnits {
 		if weight > 0 && r.numberOfEndpoints(key, port) > 0 {
 			serviceUnitNames[key] = 1
 		} else {
 			serviceUnitNames[key] = 0
 		}
 	}
       return serviceUnitNames
}
```

This logic doesn't work for the scenarios where we have > 1 service, but only 1 active service. An example:
- ServiceA has weight 0
- ServiceB has weight 256

which needs to become:
- ServiceA stays weight 0
- ServiceB becomes weight 1

Total number of services = 2. Total number of active services = 1. The suggested logic doesn't trigger because total number of services = 2.

Do that make sense?

Reply:
Is there any way to do this without iterating over the service units twice?  Maybe keep a record of the key/s of active service/s?  If there is only one key and it passes the condition, then set it.

Reply:
Hmm. Yes - I made an improvement to not iterate twice, by creating a data structure to store the "reduced" weight at the same iteration that we count in.

Then, if our counter == 1, just return the data structure with the reduced values in it. 

Should be slightly more efficient. Let me know what you think.

Reply:
> Should be slightly more efficient

I wonder how often the number of service units is above 10, 100, 1000? Is there a real-world optimisation here?

Reply:
We actually only allow a total of 4 services (1 `spec.to` and 3 `spec.alternativeBackends`).

However, my previous logic was calling `numberOfEndpoints` twice. It looks like the max number of endpoints if [1000](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#:~:text=By%20default%2C%20the%20control%20plane,to%20a%20maximum%20of%201000.).

So the last improvement went from 2 numberOfEndpoints x 4 services x 1000 endpoints to 4 services x 1000 endpoints. A minor improvement in the real-world.

Reply:
The extra allocation + GC is probably going to cost significantly more than looping twice.

Reply:
Right. I didn't think about the time it takes to allocate the list.

I don't want to over complicate...but I feel obligated to provide data vs. just trying to guess at what is more efficient. So I ran an experiment. I change the unit test `TestCalculateServiceWeights` to run this function 10 million times for each unit test case:
```
	var endpointWeights map[ServiceUnitKey]int32
	for i := 1; i <= 10000000; i++ {
		endpointWeights = router.calculateServiceWeights(tc.serviceWeights, tc.routePort)
	}
```

**Findings with my current revision (2 makes 1 iteration) compared to my previous revision (1 make 2 iterations) :**
- In the cases that there is 1 active service, current revision is ~25% faster (due to not having to count endpoints again)
- In the cases that there > 1 active service, current revision ~10% slower (due to the extra allocation)

Disclaimer that these are plug-and-chug numbers. But I think the outcome is whether we favor scenarios where there is 1 active service or > 1 active service.

If I took a wild guess, I'd say our customers are using a 50 / 50 split between 1 active service and > 1 active service routes. I can't say that either one is more common than the other. Let me know if you disagree.

I'm going to leave it as it is currently to favor 1 active service. But feel free to disagree.

Reply:
Oh but I did move the `serviceUnitNames := make(map[ServiceUnitKey]int32)` to below the active service unit logic because that wasn't needed yet, and in the case of 1 active service unit, it's an unnecessary allocation.

https://github.com/openshift/router/compare/28515b833a28db1e14853ff0044190339b02212b..3875a5660ae1d96ecce0e7325f36f35d8d0b8f1b

Reply:
You can just keep the one map serviceUnitNames and use it again later.  in the subsequent code it doesn't look at the values that are already in serviceUnitNames, it just resets the value for each key in L1428.

Reply:
True, that does work. As long as that doesn't make the code too fragile, I suppose that safe to do. Done.

Reply:
With all the micro optimisation effort ... whose dreams come true if we analyse at this level? How many milliseconds did we save (who cares)? Will anybody notice?